Account select_from(model) to query models

bodik opened the pull request:
When used with https://github.com/Pegase745/sqlalchemy-datatables/ the query is constructed with .select_from(). This PR adds ability to account entity added to query. Note that patch works on all sqla versions, but coverage had to be tuned by pragma to pass 100% constraint.

bodik committed:
Account select_from(model) to query models

bodik commented:
rebased to master

bodik commented:
amended commiter

bodik commented:
hi, any chance to have this reviewed or even merged ?

juliotrigo commented on sqlalchemy_filters/models.py:
```diff
@@ -30,6 +31,11 @@ def get_query_models(query):
         A dictionary with all the models included in the query.
     """
     models = [col_desc['entity'] for col_desc in query.column_descriptions]
+    if query._select_from_entity and (query._select_from_entity not in models):
```
`_select_from_entity` is a "private" attribute, part of the internal SQLAlchemy implementation, and it could potentially change or disappear without any further notice. I suggest we perform the following check instead to avoid accessing an attribute that does not exist:

```python
if hasattr(query, '_select_from_entity') and (query._select_from_entity not in models):
```

bodik replied:
fixed with 228567d

juliotrigo commented on sqlalchemy_filters/models.py:
```diff
@@ -30,6 +31,11 @@ def get_query_models(query):
     """
     models = [col_desc['entity'] for col_desc in query.column_descriptions]
+    if query._select_from_entity and (query._select_from_entity not in models):
+        if isinstance(query._select_from_entity, Mapper):  # pragma: no cover
```
It would be good to cover all the paths in the code and make no use of `# pragma: no cover`. This way, the 100% coverage that we have in the tests is real and we can be sure that the code does what we expect since it's covered in the test.

Another advantage of doing this is that the tests serve as documentation for developers and it's easier to understand what the code does and how to reproduce it by just looking at the tests.

Could you please try to remove the 2 `# pragma: no cover` comments and have tests for those codepaths as well?

bodik replied:
I don't think that it is possible. The conditional blocks solves the API change between sqla 1.0 and sqla 1.1+. I understand that accessing internals such as this is not so nice technique, but I've found no other way to support `.select_from` query construction.

```
(Pdb) sqlalchemy.__version__
'1.0.19'
(Pdb) type(query._select_from_entity)
<class 'sqlalchemy.ext.declarative.api.DeclarativeMeta'>
(Pdb) query._select_from_entity
<class 'test.models.Bar'>

(Pdb) sqlalchemy.__version__
'1.3.15'
(Pdb) type(query._select_from_entity)
<class 'sqlalchemy.orm.mapper.Mapper'>
(Pdb) query._select_from_entity
<Mapper at 0x7f767ec1ddd8; Bar>
(Pdb) query._select_from_entity.class_
<class 'test.models.Bar'>
```

juliotrigo replied:
It makes sense. Thanks for including the example.

juliotrigo reviewed (commented):
Thank you for all your contributions @bodik and apologies for my late review!

I have taken a look and have left a few comments.

Additionally, could you also please amend the repository documentation (`README.rst`) adding some information about what this change allows us to do? This is where we keep the different features and APIs that this library has.

bodik commented:
based on comments I'have done so far:
* rebased to master
* fixed the code accessing sqla internals
* amended the documentation
* added comments to justify the usage of `no cover` pragmas

juliotrigo commented on sqlalchemy_filters/models.py:
```diff
@@ -30,6 +31,17 @@ def get_query_models(query):
+        (query._select_from_entity not in models)
+    ):
+        if isinstance(query._select_from_entity, Mapper):  # pragma: no cover
+            # sqlalchemy>=1.1
```
👍 Thanks for the comments here. Very useful!

juliotrigo commented on sqlalchemy_filters/models.py:
```diff
@@ -30,6 +31,17 @@ def get_query_models(query):
+            models.append(query._select_from_entity.class_)
+        else:  # pragma: no cover
+            # sqlalchemy==1.0
+            models.append(query._select_from_entity)
```
Could the `query._select_from_entity` model be already present in the `models` list?

`models` has been generated from `query.column_descriptions` and we could potentially check that `type(query._select_from_entity)` is not the type of any of the objects already present in `models`.

I am just trying to avoid adding a duplicate model to the `models` list, in case that's possible.

bodik replied:
sure. i'm really sorry to waste you time, whole block was indeed wrong because the presence check was done only by sqla 1.1+ API.

juliotrigo reviewed (commented):
Thanks for addressing my comments and amending the documentation.

I have one extra question.

bodik commented:
Based on your comments (my apologies again ;), I have fully rephrased the block to add model classes attached to query by `select_from` clause. It should do proper check for both sqla dialects (1.0 vs 1.1+) now and also rephrasing allowed both `pragma: no cover` to go away. I've also squashed the PR.

juliotrigo commented on sqlalchemy_filters/models.py:
```diff
@@ -31,6 +32,20 @@ def get_query_models(query):
     models = [col_desc['entity'] for col_desc in query.column_descriptions]
     models.extend(mapper.class_ for mapper in query._join_entities)
+
+    # account also query.select_from entities
```
Nice to also have this comment here.

juliotrigo reviewed (approved):
Thanks for the new changes and no worries at all!

I also missed that the check was wrong.

The code looks good to me 👍 and it's really good that we also managed to get rid of the `pragma: no cover` comments.
